### PR TITLE
[filters] move towards index_t and Indices

### PR DESCRIPTION
**Pull request by mvieth**
See also commit messages. This is only one step, there is more to be done in filters (in another PR).

### CODE REVIEW

File: filters/include/pcl/filters/impl/extract_indices.hpp [hunk 1]
```diff
@@ -55,9 +55,9 @@ pcl::ExtractIndices<PointT>::filterDirectly (PointCloudPtr &cloud)
 
   std::vector<pcl::PCLPointField> fields;
   pcl::for_each_type<FieldList> (pcl::detail::FieldAdder<PointT> (fields));
-  for (int rii = 0; rii < static_cast<int> (removed_indices_->size ()); ++rii)  // rii = removed indices iterator
+  for (const auto& rii : (*removed_indices_)) // rii = removed indices iterator
   {
-    std::size_t pt_index = (std::size_t) (*removed_indices_)[rii];
+    std::size_t pt_index = (std::size_t) rii;
```
Comment (larshg) [inline filters/include/pcl/filters/impl/extract_indices.hpp:60]: Should be `uindex_t`?

Comment (mvieth) [reply filters/include/pcl/filters/impl/extract_indices.hpp]: Yes, that makes a bit more sense. Unsigned to avoid sign compare warnings. Changed it.

File: filters/include/pcl/filters/impl/extract_indices.hpp [hunk 2]
```diff
@@ -55,9 +55,9 @@ pcl::ExtractIndices<PointT>::filterDirectly (PointCloudPtr &cloud)
 
   std::vector<pcl::PCLPointField> fields;
   pcl::for_each_type<FieldList> (pcl::detail::FieldAdder<PointT> (fields));
-  for (int rii = 0; rii < static_cast<int> (removed_indices_->size ()); ++rii)  // rii = removed indices iterator
+  for (const auto& rii : (*removed_indices_)) // rii = removed indices iterator
   {
-    std::size_t pt_index = (std::size_t) (*removed_indices_)[rii];
+    uindex_t pt_index = (uindex_t) rii;
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/extract_indices.hpp:60]: `const auto` would be better, even preventing comparison between unsigned and signed in line 61, right?

File: filters/include/pcl/filters/impl/grid_minimum.hpp [hunk 3]
```diff
@@ -117,21 +117,21 @@ pcl::GridMinimum<PointT>::applyFilterIndices (std::vector<int> &indices)
...
   // same point of resulting CloudPoint
-  for (std::vector<int>::const_iterator it = indices_->begin (); it != indices_->end (); ++it)
+  for (const auto& it : (*indices_))
   {
     if (!input_->is_dense)
       // Check if the point is invalid
-      if (!std::isfinite ((*input_)[*it].x) ||
-          !std::isfinite ((*input_)[*it].y) ||
-          !std::isfinite ((*input_)[*it].z))
+      if (!std::isfinite ((*input_)[it].x) ||
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/grid_minimum.hpp:124]: `isXYZFinite((*input_)[it])`

File: filters/include/pcl/filters/impl/local_maximum.hpp [hunk 4]
```diff
@@ -110,28 +110,28 @@ pcl::LocalMaximum<PointT>::applyFilterIndices (std::vector<int> &indices)
   // Find all points within xy radius (i.e., a vertical cylinder) of the query
   // point, removing those that are locally maximal (i.e., highest z within the
   // cylinder)
-  for (int iii = 0; iii < static_cast<int> (indices_->size ()); ++iii)
+  for (const auto& iii : (*indices_))
   {
-    if (!isFinite ((*input_)[(*indices_)[iii]]))
+    if (!isFinite ((*input_)[iii]))
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/local_maximum.hpp:115]: `isXYZFinite((*input_)[iii])`  (explicit instead of the implicit version)

And same in other places. This can go in a separate PR, but since these lines are already being touched, I'm flagging them here

File: filters/include/pcl/filters/impl/plane_clipper3D.hpp [hunk 5]
```diff
@@ -219,9 +219,9 @@ pcl::PlaneClipper3D<PointT>::clipPointCloud3D (const pcl::PointCloud<PointT>& cl
   }
   else
   {
-    for (std::vector<int>::const_iterator iIt = indices.begin (); iIt != indices.end (); ++iIt)
-      if (clipPoint3D (cloud_in[*iIt]))
-        clipped.push_back (*iIt);
+    for (const auto& iIt : indices)
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/plane_clipper3D.hpp:222]: `It` is for iterator. A variable name change wouldn't be bad here (and other places)

**Review by kunaltyagi — COMMENTED** [review]
47/56

File: filters/include/pcl/filters/impl/extract_indices.hpp (hunk 2 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/extract_indices.hpp:60]: Do you mean like this `const auto pt_index = rii;`? Then `pt_index` would be exactly the same as `rii` and potentially signed.

File: filters/include/pcl/filters/impl/local_maximum.hpp (hunk 4 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/local_maximum.hpp:115]: Yes, that is best left for another PR, since it is not really related

File: filters/include/pcl/filters/impl/extract_indices.hpp (hunk 2 above)
Comment (kunaltyagi) [reply filters/include/pcl/filters/impl/extract_indices.hpp:60]: Ah, I forgot that the PR to make `pointcloud.size()` return `index_t` is not mature yet.

File: filters/include/pcl/filters/impl/extract_indices.hpp [hunk 6]
```diff
@@ -147,12 +147,12 @@ pcl::ExtractIndices<PointT>::applyFilterIndices (std::vector<int> &indices)
   else  // Inverted functionality
   {
     // Set up the full indices set
-    std::vector<int> full_indices (input_->size ());
-    for (int fii = 0; fii < static_cast<int> (full_indices.size ()); ++fii)  // fii = full indices iterator
+    Indices full_indices (input_->size ());
+    for (index_t fii = 0; fii < static_cast<index_t> (full_indices.size ()); ++fii)  // fii = full indices iterator
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/extract_indices.hpp:151]: `std::iota`? (and in L135)

File: filters/include/pcl/filters/impl/grid_minimum.hpp [hunk 7]
```diff
@@ -117,21 +118,19 @@ pcl::GridMinimum<PointT>::applyFilterIndices (std::vector<int> &indices)
...
 
-    int ijk0 = static_cast<int> (std::floor ((*input_)[*it].x * inverse_resolution_) - static_cast<float> (min_b[0]));
-    int ijk1 = static_cast<int> (std::floor ((*input_)[*it].y * inverse_resolution_) - static_cast<float> (min_b[1]));
+    int ijk0 = static_cast<int> (std::floor ((*input_)[index].x * inverse_resolution_) - static_cast<float> (min_b[0]));
+    int ijk1 = static_cast<int> (std::floor ((*input_)[index].y * inverse_resolution_) - static_cast<float> (min_b[1]));
 
     // Compute the grid cell index
     int idx = ijk0 * divb_mul[0] + ijk1 * divb_mul[1];
-    index_vector.emplace_back(static_cast<unsigned int> (idx), *it);
+    index_vector.emplace_back(static_cast<unsigned int> (idx), index);
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/grid_minimum.hpp:133]: I think modification of `ijk{0,1}`, `idx` and `point_index_idx` is also needed, else you're causing an overflow in the first ones and then other variables just have garbage in them (for index_t == int64_t)

File: filters/include/pcl/filters/impl/radius_outlier_removal.hpp [hunk 8]
```diff
@@ -80,10 +80,10 @@ pcl::RadiusOutlierRemoval<PointT>::applyFilterIndices (std::vector<int> &indices
     int mean_k = min_pts_radius_ + 1;
     double nn_dists_max = search_radius_ * search_radius_;
 
-    for (std::vector<int>::const_iterator it = indices_->begin (); it != indices_->end (); ++it)
+    for (const auto& index : (*indices_))
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/impl/radius_outlier_removal.hpp:83]: (in several places) Needless brackets around *indices_

**Review by kunaltyagi — COMMENTED** [review]
I think we need some PRs dedicated to move from `std::size_t` back to `index_t`

File: filters/include/pcl/filters/impl/radius_outlier_removal.hpp (hunk 8 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/radius_outlier_removal.hpp:83]: Yes, they are not necessary, but I added them for clarity and safety.

File: filters/include/pcl/filters/impl/extract_indices.hpp (hunk 6 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/extract_indices.hpp:151]: I wonder if I write `std::iota(full_indices.begin(), full_indices.end(), 0)`, what type will the 0 be? My guess is `int`. So would `std::iota(full_indices.begin(), full_indices.end(), static_cast<index_t>(0))` be sufficient?

Comment (kunaltyagi) [reply filters/include/pcl/filters/impl/extract_indices.hpp:151]: Yeah. The 2nd form is needed

Comment (mvieth) [reply filters/include/pcl/filters/impl/extract_indices.hpp]: Done

File: filters/include/pcl/filters/impl/grid_minimum.hpp (hunk 7 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/grid_minimum.hpp:133]: Maybe we can leave that for the future? Like I wrote above, this PR is only one step, to do the most obvious changes. I think a detailed analysis of types is needed in grid_minimum.hpp (and other files), since there is a lot of casting between `int`, `unsigned int`, and `float`.

**Review by kunaltyagi — APPROVED** [review]
LGTM

**Review by larshg — APPROVED** [review]
